File: sfa_api/sfa_api/Features/Billings/Entities/BillingItem.cs
using sfa_api.Features.Billings.Enums;
using sfa_api.Features.PricingStructures.Entities;
using sfa_api.Features.Products.Entities;

namespace sfa_api.Features.Billings.Entities;

public class BillingItem
{
    public int Id { get; set; }
    public int BillingId { get; set; }
    public int ProductId { get; set; }
    public decimal Quantity { get; set; }
    public decimal UnitPrice { get; set; }
    public decimal DiscountRate { get; set; }
    public decimal DiscountAmount { get; set; }
    public decimal TotalPrice { get; set; }
    public BillingItemType BillingItemType { get; set; } = BillingItemType.Sale;
    public ReturnType? ReturnType { get; set; }
    public FreeIssueSource? FreeIssueSource { get; set; }
    public DateOnly? ExpireDate { get; set; }
    public int LineNumber { get; set; }

    /// <summary>Who this line came from — the rep's original bill, or a distributor review reduction.</summary>
    public BillingItemSource Source { get; set; } = BillingItemSource.SalesRep;

    /// <summary>
    /// On a DistributorReturn line, the Sale/FreeIssue line the returned quantity was carved out of.
    /// Null on rep-entered lines.
    /// </summary>
    public int? SourceBillingItemId { get; set; }

    /// <summary>
    /// The quantity the rep originally billed, captured the first time the distributor reduces this
    /// line. Null while the line has never been adjusted. Repeat adjustments do not overwrite it.
    /// </summary>
    public decimal? OriginalQuantity { get; set; }

    /// <summary>
    /// The pricing structure that priced this line. Null on lines created before pricing
    /// structures existed. Historical: later edits to the structure never touch this line —
    /// <see cref="UnitPrice"/> and <see cref="ListUnitPrice"/> are the frozen prices.
    /// </summary>
    public int? PricingStructureId { get; set; }

    /// <summary>Which structure price the line used (pack, case, or a rep-typed return price). Null on legacy lines.</summary>
    public PriceBasis? PriceBasis { get; set; }

    /// <summary>
    /// The structure's price for <see cref="PriceBasis"/> at billing time — the exact case price on a
    /// Case line (whose <see cref="UnitPrice"/> is the per-pack equivalent). Null on Manual and legacy lines.
    /// </summary>
    public decimal? ListUnitPrice { get; set; }

    // Audit
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public bool IsDeleted { get; set; } = false;

    // Navigation
    public Billing Billing { get; set; } = null!;
    public Product Product { get; set; } = null!;
    public BillingItem? SourceBillingItem { get; set; }
    public PricingStructure? PricingStructure { get; set; }
}

File: sfa_mobile/lib/features/supervisor_billing/domain/entities/billing_item.dart
import 'package:equatable/equatable.dart';

enum BillingItemType { sale, returnItem }

enum ReturnType { marketResell, damage, expire }

class BillingItem extends Equatable {
  final int id;
  final int productId;
  final String productCode;
  final String productDescription;
  final double quantity;
  final double unitPrice;
  final double discountRate;
  final double discountAmount;
  final double totalPrice;
  final bool isFreeIssue;
  final BillingItemType billingItemType;
  final ReturnType? returnType;
  final int lineNumber;

  const BillingItem({
    required this.id,
    required this.productId,
    required this.productCode,
    required this.productDescription,
    required this.quantity,
    required this.unitPrice,
    required this.discountRate,
    required this.discountAmount,
    required this.totalPrice,
    required this.isFreeIssue,
    required this.billingItemType,
    this.returnType,
    required this.lineNumber,
  });

  @override
  List<Object?> get props => [id];
}
